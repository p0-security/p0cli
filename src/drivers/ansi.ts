/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { mapValues } from "lodash";

const AnsiCodes = {
  Reset: "00",
  Dim: "02",
  Green: "32",
  Yellow: "33",
} as const;

export const Ansi = (value: string) => `\u001b[${value}`;

/** Creates an ANSI Select Graphic Rendition code */
export const AnsiSgr = mapValues(AnsiCodes, (v) => Ansi(v + "m"));
