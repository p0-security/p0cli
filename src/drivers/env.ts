/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import dotenv from "dotenv";

dotenv.config();

const { env } = process;

export const config = {
  fs: {
    // Falls back to public production Firestore credentials
    apiKey: env.P0_FS_API_KEY ?? "AIzaSyCaL-Ik_l_5tdmgNUNZ4Nv6NuR4o5_PPfs",
    authDomain: env.P0_FS_AUTH_DOMAIN ?? "p0-prod.firebaseapp.com",
    projectId: env.P0_FS_PROJECT_ID ?? "p0-prod",
    storageBucket: env.P0_FS_STORAGE_BUCKET ?? "p0-prod.appspot.com",
    messagingSenderId: env.P0_FS_MESSAGING_SENDER_ID ?? "228132571547",
    appId: env.P0_FS_APP_ID ?? "1:228132571547:web:4da03aeb78add86fe6b93e",
  },
  google: {
    clientId:
      env.P0_GOOGLE_OIDC_CLIENT_ID ??
      "228132571547-kilcq1er15hlbl6mitghttnacp7u58l8.apps.googleusercontent.com",
    // Despite the name, this is not actually "secret" in any sense of the word.
    // Instead, the client is protected by requiring PKCE and defining the redirect URIs.
    clientSecret:
      env.P0_GOOGLE_OIDC_CLIENT_SECRET ?? "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",
  },
  appUrl: env.P0_APP_URL ?? "https://api.p0.app",
  ENV: env.P0_ENV ?? "production",
};
