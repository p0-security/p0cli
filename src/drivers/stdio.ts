/** Copyright Â© 2024-present P0 Security 

This file is part of @p0security/p0cli

@p0security/p0cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/p0cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/p0cli. If not, see <https://www.gnu.org/licenses/>.
**/

/** Functions to handle stdio
 *
 * These are essentially wrappers around console.foo, but allow for
 * - Better testing
 * - Later redirection / duplication
 */

/** Used to output machine-readable text to stdout
 *
 * In general this should not be used for text meant to be consumed
 * only by humans.
 */
export function print1(message: any) {
  // eslint-disable-next-line no-console
  console.log(message);
}

/** Output human-consumable text to stderr
 *
 * In general this should not be used for machine-consumed text.
 */
export function print2(message: any) {
  // eslint-disable-next-line no-console
  console.error(message);
}
