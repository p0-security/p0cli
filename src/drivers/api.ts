/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { Authn } from "../types/identity";
import { p0VersionInfo } from "../version";
import { getTenantConfig } from "./config";
import { defaultConfig } from "./env";
import { print2 } from "./stdio";
import * as path from "node:path";
import yargs from "yargs";

const DEFAULT_PERMISSION_REQUEST_TIMEOUT = 300e3; // 5 minutes

const tenantOrgUrl = (tenant: string) =>
  `${getTenantConfig()?.appUrl ?? defaultConfig.appUrl}/orgs/${tenant}`;
const tenantUrl = (tenant: string) => `${getTenantConfig().appUrl}/o/${tenant}`;
const publicKeysUrl = (tenant: string) =>
  `${tenantUrl(tenant)}/integrations/ssh/public-keys`;
const sshAuditUrl = (tenant: string) =>
  `${tenantUrl(tenant)}/integrations/ssh/audit`;

const commandUrl = (tenant: string) => `${tenantUrl(tenant)}/command/`;
const adminLsCommandUrl = (tenant: string) => `${tenantUrl(tenant)}/command/ls`;
export const tracesUrl = (tenant: string) => `${tenantUrl(tenant)}/traces`;

export const fetchOrgData = async <T>(orgId: string) =>
  unauthenticatedApiFetch<T>(tenantOrgUrl(orgId), "GET");

export const fetchAccountInfo = async <T>(authn: Authn) =>
  baseFetch<T>(authn, {
    url: `${tenantUrl(authn.identity.org.slug)}/account`,
    method: "GET",
  });

export const fetchPermissionRequestDetails = async <T>(
  authn: Authn,
  requestId: string
) =>
  baseFetch<T>(authn, {
    url: `${tenantUrl(authn.identity.org.slug)}/permission-requests/${requestId}`,
    method: "GET",
    maxTimeoutMs: DEFAULT_PERMISSION_REQUEST_TIMEOUT,
  });

export const fetchIntegrationConfig = async <T>(
  authn: Authn,
  integration: string
) =>
  baseFetch<T>(authn, {
    url: `${tenantUrl(authn.identity.org.slug)}/integrations/${integration}/config`,
    method: "GET",
  });

export const fetchStreamingCommand = async function* <T>(
  authn: Authn,
  args: yargs.ArgumentsCamelCase,
  argv: string[]
) {
  yield* fetchWithStreaming<T>(authn, {
    url: commandUrl(authn.identity.org.slug),
    method: "POST",
    body: JSON.stringify({
      argv,
      scriptName: path.basename(args.$0),
      wait: true,
    }),
  });
};

export const fetchCommand = async <T>(
  authn: Authn,
  args: yargs.ArgumentsCamelCase,
  argv: string[]
) =>
  baseFetch<T>(authn, {
    url: commandUrl(authn.identity.org.slug),
    method: "POST",
    body: JSON.stringify({
      argv,
      scriptName: path.basename(args.$0),
    }),
  });

/** Special admin 'ls' command that can retrieve results for all users. Requires 'owner' permission. */
export const fetchAdminLsCommand = async <T>(
  authn: Authn,
  args: yargs.ArgumentsCamelCase,
  argv: string[]
) =>
  baseFetch<T>(authn, {
    url: adminLsCommandUrl(authn.identity.org.slug),
    method: "POST",
    body: JSON.stringify({
      argv,
      scriptName: path.basename(args.$0),
    }),
  });

export const submitPublicKey = async <T>(
  authn: Authn,
  args: { publicKey: string; requestId: string }
) =>
  baseFetch<T>(authn, {
    url: publicKeysUrl(authn.identity.org.slug),
    method: "POST",
    body: JSON.stringify({
      requestId: args.requestId,
      publicKey: args.publicKey,
    }),
  });

export const fetchWithStreaming = async function* <T>(
  authn: Authn,
  args: {
    url: string;
    method: string;
    body?: string;
    maxTimeoutMs?: number;
  }
) {
  const token = await authn.getToken();
  const { version } = p0VersionInfo;
  const { url, method, body, maxTimeoutMs } = args;
  const fetchOptions = {
    method,
    headers: {
      authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      "User-Agent": `P0 CLI/${version}`,
    },
    body,
    keepalive: true,
  };
  const textDecoder = new TextDecoder();
  try {
    const response = await fetch(
      url,
      maxTimeoutMs
        ? { ...fetchOptions, signal: AbortSignal.timeout(maxTimeoutMs) }
        : fetchOptions
    );
    const reader = response.body?.getReader();
    if (!reader) throw `No reader available`;
    while (true) {
      const read = await reader.read();
      if (read.done) {
        break;
      }
      const value = read.value;
      const text = textDecoder.decode(value);
      const jsonlSegments = text.split("\n").filter(Boolean);
      for (const segment of jsonlSegments) {
        const parsedResponse = JSON.parse(segment);
        if (parsedResponse.type === "error") {
          throw parsedResponse.error;
        }
        if (parsedResponse.type === "heartbeat") {
          continue;
        }
        if (parsedResponse.type !== "data" || !("data" in parsedResponse)) {
          throw "Invalid response from the server";
        }
        const { data } = parsedResponse;
        if ("error" in data) {
          throw data.error;
        }
        yield data as T;
      }
    }
  } catch (error) {
    if (
      error instanceof TypeError &&
      (error.message === "fetch failed" || error.message === "terminated")
    ) {
      throw `Network error: Unable to reach the server.`;
    } else {
      throw error;
    }
  }
};

export const auditSshSessionActivity = async (args: {
  authn: Authn;
  requestId: string;
  sshSessionId: string;
  action: `ssh.session.${"end" | "start"}`;
  debug: boolean | undefined;
}) => {
  const { authn, requestId, action, sshSessionId, debug } = args;

  if (debug) {
    print2(
      `Submitting audit log for request: ${requestId}, action: ${action}, sshSessionId: ${sshSessionId}`
    );
  }

  try {
    await baseFetch(authn, {
      url: sshAuditUrl(authn.identity.org.slug),
      method: "POST",
      body: JSON.stringify({
        requestId,
        action,
        sshSessionId,
      }),
    });
    if (debug) {
      print2(`Audit log submitted for request: ${requestId}`);
    }
  } catch (error) {
    if (debug) {
      print2(`Failed to submit audit log for request: ${requestId}`);
      print2(`Error: ${JSON.stringify(error)}`);
    }
  }
};

export const baseFetch = async <T>(
  authn: Authn,
  args: {
    url: string;
    method: string;
    body?: string;
    maxTimeoutMs?: number;
  }
) => {
  const { version } = p0VersionInfo;
  const { url, method, body, maxTimeoutMs } = args;
  const token = await authn.getToken();
  const fetchOptions = {
    method,
    headers: {
      authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      "User-Agent": `P0 CLI/${version}`,
    },
    body,
    keepalive: true,
  };

  try {
    const response = await fetch(
      url,
      maxTimeoutMs
        ? { ...fetchOptions, signal: AbortSignal.timeout(maxTimeoutMs) }
        : fetchOptions
    );
    const text = await response.text();
    const data = JSON.parse(text);
    if ("error" in data) {
      throw data.error;
    }
    return data as T;
  } catch (error) {
    if (error instanceof TypeError && error.message === "fetch failed") {
      throw `Network error: Unable to reach the server at ${url}.`;
    } else {
      throw error;
    }
  }
};

const unauthenticatedApiFetch = async <T>(
  url: string,
  method: string,
  body?: string
) => {
  const { version } = p0VersionInfo;
  try {
    const fetchConfig = {
      method,
      headers: {
        "Content-Type": "application/json",
        "User-Agent": `P0 CLI/${version}`,
      },
    };
    const response = await fetch(url, {
      ...fetchConfig,
      body,
    });
    const text = await response.text();
    const data = JSON.parse(text);
    if ("error" in data) {
      throw data.error;
    }
    return data as T;
  } catch (error) {
    if (error instanceof TypeError && error.message === "fetch failed") {
      throw `Network error: Unable to reach the server at ${url}.`;
    } else {
      throw error;
    }
  }
};
