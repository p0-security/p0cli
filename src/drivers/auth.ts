import { login } from "../commands/login";
import { Identity } from "../types/identity";
import { auth } from "./firestore";
import {
  OAuthProvider,
  SignInMethod,
  signInWithCredential,
} from "firebase/auth";
import * as fs from "fs/promises";
import * as os from "os";
import * as path from "path";

export const IDENTITY_FILE_PATH = path.join(
  os.homedir(),
  ".p0",
  "identity.json"
);

export const loadCredentials = async (options?: { noRefresh?: boolean }) => {
  try {
    const buffer = await fs.readFile(IDENTITY_FILE_PATH);
    const identity: Identity = JSON.parse(buffer.toString());
    if (
      !options?.noRefresh &&
      identity.credential.expires_at < Date.now() * 1e-3
    ) {
      await login({ org: identity.org.slug }, { skipAuthenticate: true });
      console.error("\u200B"); // Force a new line
    }
    return identity;
  } catch (error: any) {
    if (error?.code === "ENOENT") {
      throw "Please run `p0 login <organization>` to use the P0 CLI.";
    }
    throw error;
  }
};

export const authenticate = async (options?: { noRefresh?: boolean }) => {
  const identity = await loadCredentials(options);
  const { credential } = identity;
  // TODO: Move to map lookup
  const provider = new OAuthProvider(
    identity.org.ssoProvider === "google"
      ? SignInMethod.GOOGLE
      : identity.org.providerId
  );
  const firebaseCredential = provider.credential({
    accessToken: credential.access_token,
    idToken: credential.id_token,
  });
  auth.tenantId = identity.org.tenantId;
  const userCredential = await signInWithCredential(auth, firebaseCredential);
  return { userCredential, identity };
};
