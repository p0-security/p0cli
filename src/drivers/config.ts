/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { Config } from "../types/org";
import { getConfigFilePath } from "./auth/path";
import { defaultConfig } from "./env";
import { getOrgData } from "./org";
import { print2 } from "./stdio";
import fs from "fs/promises";
import path from "path";

let tenantConfig: Config;

export const getTenantConfig = () => tenantConfig;

export const getContactMessage = () =>
  tenantConfig?.contactMessage ?? defaultConfig.contactMessage;

export const getHelpMessage = () =>
  tenantConfig?.helpMessage ?? defaultConfig.helpMessage;

export const shouldSkipCheckVersion = () =>
  defaultConfig.skipVersionCheck === "true" ? true : false;

/** Use only if the organization is configured with Google login to P0 */
export const getGoogleTenantConfig = () => {
  if ("google" in tenantConfig) {
    return tenantConfig;
  }
  throw `Login failed!\nThis organization is configured to use Google login but the required OAuth client parameters are missing.\n${getContactMessage()}`;
};

export const saveConfig = async (orgId: string, debug?: boolean) => {
  const orgData = await getOrgData(orgId);

  if (debug) {
    print2(`Retrieved org data: ${JSON.stringify(orgData)}`);
  }

  const config = orgData.config ?? defaultConfig;

  const configFilePath = getConfigFilePath();

  print2(`Saving config to ${configFilePath}.`);

  const dir = path.dirname(configFilePath);
  await fs.mkdir(dir, { recursive: true });
  await fs.writeFile(configFilePath, JSON.stringify(config), { mode: "600" });

  tenantConfig = config;
};

export const loadConfig = async () => {
  const buffer = await fs.readFile(getConfigFilePath());
  tenantConfig = JSON.parse(buffer.toString());
  return tenantConfig;
};
