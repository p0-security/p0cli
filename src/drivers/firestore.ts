/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { Identity } from "../types/identity";
import { getContactMessage, loadConfig } from "./config";
import { bootstrapConfig } from "./env";
import { print2 } from "./stdio";
import { EXPIRED_CREDENTIALS_MESSAGE } from "./util";
import { FirebaseApp, FirebaseError, initializeApp } from "firebase/app";
import {
  getAuth,
  OAuthProvider,
  SignInMethod,
  signInWithCredential,
  UserCredential,
} from "firebase/auth";
import {
  collection as fsCollection,
  CollectionReference,
  doc as fsDoc,
  DocumentReference,
  getFirestore,
  terminate,
  Firestore,
} from "firebase/firestore";

const bootstrapApp = initializeApp(bootstrapConfig.fs, "bootstrapApp");
const bootstrapFirestore = getFirestore(bootstrapApp);

let app: FirebaseApp;
let firestore: Firestore;

export async function initializeFirebase() {
  if (!firestore) {
    const tenantConfig = await loadConfig();
    app = initializeApp(tenantConfig.fs, "authFirebase");
    firestore = getFirestore(app);
  }
}

const findProviderId = (identity: Identity) => {
  switch (identity.org.ssoProvider) {
    case "google":
      return SignInMethod.GOOGLE;
    case "google-oidc":
      return "oidc.google-oidc";
    default:
      return identity.org.providerId;
  }
};

export async function authenticateToFirebase(
  identity: Identity,
  options?: {
    debug?: boolean;
  }
): Promise<UserCredential> {
  const { credential } = identity;
  const tenantId = identity.org.tenantId;

  await initializeFirebase();

  const provider = new OAuthProvider(findProviderId(identity));

  const firebaseCredential = provider.credential({
    accessToken: credential.access_token,
    idToken: credential.id_token,
  });

  const auth = getAuth(app);
  auth.tenantId = tenantId;

  let userCredential;
  try {
    userCredential = await signInWithCredential(auth, firebaseCredential);
  } catch (error) {
    if (
      error instanceof FirebaseError &&
      error.code === "auth/invalid-credential"
    ) {
      throw EXPIRED_CREDENTIALS_MESSAGE;
    } else {
      if (options?.debug) {
        print2("Authentication error: " + error);
      }
      throw `An unexpected error occurred during authentication.\n${getContactMessage()}`;
    }
  }

  if (!userCredential?.user?.email) {
    throw `Can not sign in: this user has previously signed in with a different identity provider.\n${getContactMessage()}`;
  }

  return userCredential;
}

export const collection = <T>(path: string, ...pathSegments: string[]) => {
  return fsCollection(
    firestore,
    path,
    ...pathSegments
  ) as CollectionReference<T>;
};

export const doc = <T>(path: string) => {
  return fsDoc(firestore, path) as DocumentReference<T>;
};

export const bootstrapDoc = <T>(path: string) => {
  return fsDoc(bootstrapFirestore, path) as DocumentReference<T>;
};

/** Ensures that Firestore is shutdown at command termination
 *
 * This prevents Firestore from holding the command on execution completion or failure.
 */
export const fsShutdownGuard =
  <P, T>(cb: (args: P) => Promise<T>) =>
  async (args: P) => {
    try {
      await cb(args);
    } finally {
      shutdownFirebase();
    }
  };

export const shutdownFirebase = () => {
  if (bootstrapFirestore) void terminate(bootstrapFirestore);
  if (firestore) void terminate(firestore);
};
