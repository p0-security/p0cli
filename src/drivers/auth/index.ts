/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { login } from "../../commands/login";
import { Authn, Identity } from "../../types/identity";
import { TokenResponse } from "../../types/oidc";
import { OrgData } from "../../types/org";
import { authenticateToFirebase } from "../firestore";
import { print2 } from "../stdio";
import { getIdentityCachePath, getIdentityFilePath } from "./path";
import * as fs from "fs/promises";
import * as os from "os";
import * as path from "path";

export const cached = async <T>(
  name: string,
  loader: () => Promise<T>,
  options: { duration: number },
  hasExpired?: (data: T) => boolean
): Promise<T> => {
  const identityCachePath = getIdentityCachePath();

  // Following lines sanitize input
  // nosemgrep: javascript.lang.security.audit.path-traversal.path-join-resolve-traversal.path-join-resolve-traversal
  const loc = path.resolve(path.join(identityCachePath, `${name}.json`));
  if (!loc.startsWith(identityCachePath)) {
    throw new Error("Illegal path traversal");
  }

  const loadCache = async () => {
    const data = await loader();
    if (!data) throw `Could not load credentials for "${name}"`;
    await fs.mkdir(path.dirname(loc), { recursive: true, mode: "700" });
    await fs.writeFile(loc, JSON.stringify(data), { mode: "600" });
    return data;
  };

  try {
    const stat = await fs.stat(loc);
    if (stat.mtime.getTime() < Date.now() - options.duration) {
      await fs.rm(loc);
      return await loadCache();
    }

    const data = JSON.parse((await fs.readFile(loc)).toString("utf-8")) as T;
    if (hasExpired?.(data)) {
      await fs.rm(loc);
      return await loadCache();
    }
    return data;
  } catch (error: any) {
    if (error?.code !== "ENOENT")
      print2(
        `Could not load credentials "${name}" from cache: ${error.message ?? error}`
      );
    return await loadCache();
  }
};

const clearIdentityFile = async () => {
  try {
    const identityFilePath = getIdentityFilePath();
    // check to see if the file exists before trying to remove it
    await fs.access(identityFilePath);
    await fs.rm(identityFilePath);
  } catch {
    return;
  }
};

const clearIdentityCache = async () => {
  try {
    const identityCachePath = getIdentityCachePath();
    // check to see if the directory exists before trying to remove it
    await fs.access(identityCachePath);
    await fs.rm(identityCachePath, { recursive: true });
  } catch {
    return;
  }
};

const loadCredentialsWithAutoLogin = async (options?: {
  noRefresh?: boolean;
}): Promise<Identity> => {
  try {
    const buffer = await fs.readFile(getIdentityFilePath());
    const identity: Identity = JSON.parse(buffer.toString());
    if (
      !options?.noRefresh &&
      identity.credential.expires_at < Date.now() * 1e-3
    ) {
      await login({ org: identity.org.slug }, { skipAuthenticate: true });
      print2("\u200B"); // Force a new line
      return loadCredentialsWithAutoLogin({ noRefresh: true });
    }
    return identity;
  } catch (error: any) {
    if (error?.code === "ENOENT") {
      throw "Please run `p0 login <organization>` to use the P0 CLI.";
    }
    throw error;
  }
};

export const writeIdentity = async (
  org: OrgData,
  credential: TokenResponse
) => {
  await clearIdentityCache();

  const identityFilePath = getIdentityFilePath();

  const expires_at = Date.now() * 1e-3 + credential.expires_in - 1; // Add 1 second safety margin
  print2(`Saving authorization to ${identityFilePath}.`);
  const dir = path.dirname(identityFilePath);
  await fs.mkdir(dir, { recursive: true });
  await fs.writeFile(
    identityFilePath,
    JSON.stringify(
      {
        credential: { ...credential, expires_at },
        org,
      },
      null,
      2
    ),
    {
      mode: "600",
    }
  );
};

export const deleteIdentity = async () => {
  await clearIdentityCache();
  await clearIdentityFile();
};

export const authenticate = async (options?: {
  noRefresh?: boolean;
}): Promise<Authn> => {
  const identity = await loadCredentialsWithAutoLogin(options);
  const userCredential = await authenticateToFirebase(identity);

  return { userCredential, identity };
};
