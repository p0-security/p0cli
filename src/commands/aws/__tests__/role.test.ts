/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { awsCommand } from "..";
import { print1, print2 } from "../../../drivers/stdio";
import { mockGetDoc } from "../../../testing/firestore";
import { failure } from "../../../testing/yargs";
import { samlResponse } from "./__input__/saml-response";
import { stsResponse } from "./__input__/sts-response";
import yargs from "yargs";

jest.mock("fs/promises");
jest.mock("../../../drivers/auth");
jest.mock("../../../drivers/stdio");
jest.mock("typescript", () => ({
  ...jest.requireActual("typescript"),
  sys: {
    writeOutputIsTTY: () => true,
  },
}));
jest.mock("../util", () => ({
  provisionRequest: jest.fn(),
}));

const mockFetch = jest.spyOn(global, "fetch");
const mockPrint1 = print1 as jest.Mock;
const mockPrint2 = print2 as jest.Mock;

beforeEach(() => {
  jest.clearAllMocks();
  mockFetch.mockImplementation(
    async (url: RequestInfo | URL) =>
      ({
        ok: true,
        // This is the token response from fetchSsoWebToken
        json: async () => ({}),
        // This is the XML response from fetchSamlResponse or stsAssumeRole
        text: async () =>
          (url as string).match(/okta.com/) ? samlResponse : stsResponse,
      }) as Response
  );
});

describe("aws role", () => {
  describe("a single installed account", () => {
    const item = {
      label: "test",
      state: "installed",
    };
    describe("without Okta SAML", () => {
      mockGetDoc({ "iam-write": { "1": item } });
      describe.each([["assume", "aws role assume Role1"]])(
        "%s",
        (_, command) => {
          it("should print a friendly error message", async () => {
            const error = await failure(awsCommand(yargs()), command);
            expect(error).toMatchInlineSnapshot(
              `"Account test is not configured for Okta SAML login."`
            );
          });
        }
      );
    });
    describe("with Okta SAML", () => {
      beforeEach(() => {
        mockGetDoc({
          "iam-write": {
            "1": {
              ...item,
              login: {
                type: "federated",
                provider: {
                  type: "okta",
                  appId: "0oabcdefgh",
                  identityProvider: "okta",
                },
              },
            },
          },
        });
      });
      describe("assume", () => {
        it("should assume a role", async () => {
          await awsCommand(yargs()).parse("aws role assume Role1");
          expect(mockPrint2.mock.calls).toMatchSnapshot("stderr");
          expect(mockPrint1.mock.calls).toMatchSnapshot("stdout");
        });
      });
    });
  });
});
