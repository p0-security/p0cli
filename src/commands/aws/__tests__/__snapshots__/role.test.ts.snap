// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aws role a single installed account with Okta SAML assume should assume a role 1`] = `
[
  [
    "Execute the following commands:
",
  ],
  [
    "
Or, populate these environment variables using BASH command substitution:

  $(p0 aws role assume undefined)
",
  ],
]
`;

exports[`aws role a single installed account with Okta SAML assume should assume a role 2`] = `
[
  [
    "  export AWS_ACCESS_KEY_ID=test-access-key
  export AWS_SECRET_ACCESS_KEY=secret-access-key
  export AWS_SESSION_TOKEN=session-token",
  ],
]
`;

exports[`aws role a single installed account with Okta SAML ls lists roles 1`] = `
[
  [
    "Your available roles for account 1:",
  ],
]
`;

exports[`aws role a single installed account with Okta SAML ls lists roles 2`] = `
[
  [
    "  Role1
  Role2",
  ],
]
`;
